cl/corpus_importer/management/commands/update_casenames_wl_dataset.py

```diff
@@ -0,0 +1,360 @@
+import logging
+import re
+import time
+from datetime import date, datetime
+
+import pandas as pd
+from django.core.management.base import BaseCommand, CommandError
+from django.db.models import Q, QuerySet
+from eyecite import get_citations
+from eyecite.models import FullCaseCitation
+from eyecite.tokenizers import HyperscanTokenizer
+
+from cl.corpus_importer.utils import add_citations_to_cluster
+from cl.search.models import Citation, OpinionCluster
+
+logger = logging.getLogger(__name__)
+HYPERSCAN_TOKENIZER = HyperscanTokenizer(cache_dir=".hyperscan")
+
+# Compile regex pattern once for efficiency
+WORD_PATTERN = re.compile(r"\b\w+\b|\b\w+\.\b")
+
+FALSE_POSITIVES = {
+    "and",
+    "personal",
+    "restraint",
+    "matter",
+    "county",
+    "city",
+    "of",
+    "the",
+    "estate",
+    "in",
+    "inc",
+    "re",
+    "st",
+    "ex",
+    "rel",
+    "v",
+    "vs",
+    "for",
+    "a",
+}
+
+DATE_FORMATS = ("%B %d, %Y", "%d-%b-%y", "%m/%d/%Y", "%m/%d/%y", "%b. %d, %Y")
+
+
+def tokenize_case_name(case_name: str) -> list[str]:
+    """Tokenizes case name and removes single-character words except for letters with periods.
+
+    :param case_name: case name to tokenize
+    :return: list of words
+    """
+    return [
+        word.lower()
+        for word in WORD_PATTERN.findall(case_name)
+        if len(word) > 1
+    ]
+
+
+def check_case_names_match(csv_case_name: str, cl_case_name: str) -> bool:
+    """Compare two case name and decide whether they are the same or not
+
+    :param csv_case_name: case name from csv
+    :param cl_case_name: case name from cluster
+    :return: True if they match else False
+    """
+    # Tokenize each string, capturing both words and abbreviations with periods and
+    # convert all words to lowercase for case-insensitive matching
+    csv_case_name_tokens = set(tokenize_case_name(csv_case_name))
+    cluster_case_name_tokens = set(tokenize_case_name(cl_case_name))
+
+    # Check if there is an overlap between case names and remove false positive words
+    overlap = csv_case_name_tokens & cluster_case_name_tokens - FALSE_POSITIVES
+
+    if not overlap:
+        # if no hits no match on name - move along
+        return False
+
+    # Check for "v." in title
+    if "v." not in csv_case_name.lower():
+        # in the matter of Smith
+        # if no V. - likely an "in re" case and only match on at least 1 name
+        return True
+
+    # otherwise check if a match occurs on both sides of the V
+    v_index = csv_case_name.lower().index("v.")
+    hit_indices = [csv_case_name.lower().find(hit) for hit in overlap]
+
+    return min(hit_indices) < v_index < max(hit_indices)
+
+
+def parse_date(date_str: str) -> date | None:
```

Comment:
shouldn't this be optional date ?

Reply:
it is suggested to use | for union types instead of Optional when working with python 3.10+ (PEP 604)